### PR TITLE
Fix sendgrid ASM config

### PR DESCRIPTION
**Pull request by RtpChris**
Currently the ASM config doesn't work with the sendgrid actions.  This is because the Sendgrid API requires the group_id and other params to be integers, not strings.

This code parses the ASM object as integers if set.  It also adds the two ASM subfields as their own parameters so that the Sendgrid API can be queried for the groups (like the templateId param)

Sendgrid API reference: https://docs.sendgrid.com/api-reference/mail-send/mail-send

### CODE REVIEW

**Comment by pipedream-component-development** [comment]
Thank you so much for submitting this! We've added it to our backlog to review, and our team has been notified.

**Comment by pipedream-component-development** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

**Comment by RtpChris** [comment]
Here's the error that we get currently when trying to use this:

```
Unexpected error (status code: 400): { "headers": { "server": "nginx", "date": "Fri, 10 Mar 2023 00:51:02 GMT", "content-type": "application/json", "content-length": "300", "connection": "close", "access-control-allow-origin": "https://sendgrid.api-docs.io", "access-control-allow-methods": "POST", "access-control-allow-headers": "Authorization, Content-Type, On-behalf-of, x-sg-elas-acl", "access-control-max-age": "600", "x-no-cors-reason": "https://sendgrid.com/docs/Classroom/Basics/API/cors.html", "strict-transport-security": "max-age=600; includeSubDomains" }, "body": { "errors": [ { "message": "Invalid type. Expected: object, given: string.", "field": "personalizations.0.bcc.0", "help": "http://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html#message.personalizations.bcc" }, { "message": "Invalid type. Expected: object, given: string.", "field": "asm", "help": null } ] } }
```

File: components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.mjs [hunk 1]
```diff
@@ -266,6 +278,45 @@ export default {
...
+        }
+        if (groupIds.length > 0) {
+          asmConfig.groups_to_display = groupIds;
+        }
+      }
+      //if asmGroupsToDisplay is configured, copy that over
+      if (this.asmGroupsToDisplay) {
+        asmConfig.groups_to_display = this.asmGroupsToDisplay;
+      }
+    }
```
Comment (GTFalcao) [inline components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.mjs:318]: This entire block of code is the same for both actions - perhaps it should be a method that is common to both of them?

e.g. a `getAsmConfig` method with this code can be defined in a common file that is imported and called by both components:
```
const asmConfig = this.getAsmConfig(); // returns undefined if none of the conditions are met
```

**Review by GTFalcao — COMMENTED** [review]
Hi @RtpChris , nice work, thanks for your contribution!

I made a comment suggesting reusing common code. Also, you should [update the patch version](https://pipedream.com/docs/components/guidelines/#versioning) for the components changed here, as well as the app's other components (since they import the `app` file, which was updated here as well).

The app's `package.json` file also needs its version increased.

Let me know if you have any questions!

File: components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.mjs (hunk 1 above)
Comment (RtpChris) [reply components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.mjs:318]: Makes sense; I'll move it to the sendgrid object.

**Review by GTFalcao — COMMENTED** [review]
Hey @RtpChris this looks very nice, thanks for the changes!

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[SendGrid_5673_2151.pdf](https://github.com/PipedreamHQ/pipedream/files/10951177/SendGrid_5673_2151.pdf)

**Comment by RtpChris** [comment]
I've added a new validator for the asmGroupsToShow, set asmGroupsToShow as optional, and have the fetch return undefined if the user doesn't have any ASMs configured

**Comment by RtpChris** [comment]
@michelle0927 - Thanks!  I think I've gotten the conflict resolved; sorry for the mess!

**Comment by vunguyenhung** [comment]
Hello @RtpChris, I'm still facing error while fetching for ASM Group ID and ASM Groups to Display



image

Would you mind checking again and sharing the steps taken to test this?

**Comment by RtpChris** [comment]
@vunguyenhung - I tested it locally as well as I could, but I'm probably missing something.  I think this time I screwed up by returning undefined instead of an int

Does the sendgrid account you're using have any suppression groups defined?  I'm still getting familiar with the pipe dream code and I'm not entirely sure how to return a custom error message for that (I based the options code on the work that was already there)

**Comment by michelle0927** [comment]
> @vunguyenhung - I tested it locally as well as I could, but I'm probably missing something. I think this time I screwed up by returning undefined instead of an int
> 
> Does the sendgrid account you're using have any suppression groups defined? I'm still getting familiar with the pipe dream code and I'm not entirely sure how to return a custom error message for that (I based the options code on the work that was already there)

@RtpChris Sorry I overlooked this during review. The prop options are expecting an array. Try returning `[]` instead of `undefined`.
```
     async options() {
        const asmGroups = await this.listAsmGroups();
        if (asmGroups) {
          return asmGroups.map((group) => ({
            label: group.name,
            value: group.id,
          }));
        }
        return [];
      },
```

**Comment by RtpChris** [comment]
> > @vunguyenhung - I tested it locally as well as I could, but I'm probably missing something. I think this time I screwed up by returning undefined instead of an int
> > Does the sendgrid account you're using have any suppression groups defined? I'm still getting familiar with the pipe dream code and I'm not entirely sure how to return a custom error message for that (I based the options code on the work that was already there)
> 
> @RtpChris Sorry I overlooked this during review. The prop options are expecting an array. Try returning `[]` instead of `undefined`.
> 
> ```
>      async options() {
>         const asmGroups = await this.listAsmGroups();
>         if (asmGroups) {
>           return asmGroups.map((group) => ({
>             label: group.name,
>             value: group.id,
>           }));
>         }
>         return [];
>       },
> ```

Ahh ok thanks!  I've made those changes

File: components/sendgrid/actions/common/common.mjs [hunk 2]
```diff
@@ -101,5 +123,52 @@ export default {
...
       return object;
     },
+    /**
+     * Prepares the asm object or the asmGroupId. Sendgrid API requires the
+     * params to be an int, so make sure it can be parsed as one
+     *
+     * @returns an object with the asm configuration, or `undefined` if none
+     */
+    getAsmConfig() {
+      //
```
Comment (vellames) [inline components/sendgrid/actions/common/common.mjs:133]: Looks like you forgot this comment here

**Review by vellames — CHANGES_REQUESTED** [review]

**Comment by dannyroosevelt** [comment]
@michelle0927 @vellames what's left here to get this merged?

**Comment by vellames** [comment]
It was just a comment left in the code, not a big deal, we can move forward with this if you want

**Comment by binamov** [comment]
can this be merged?

**Comment by dannyroosevelt** [comment]
@binamov looks like it's waiting to be QA'd -- have you gotten a chance to test the changes end to end?

**Comment by binamov** [comment]
> @binamov looks like it's waiting to be QA'd -- have you gotten a chance to test the changes end to end?

nope. I'm blocked with this integration as well, i'm trusting the numerous positive responses before mine.

**Comment by dannyroosevelt** [comment]
@vunguyenhung @michelle0927  what's the latest on this one?

**Comment by dannyroosevelt** [comment]
@michelle0927 https://github.com/PipedreamHQ/pipedream/actions/runs/4874592849/jobs/8695733981?pr=5673

**Review by michelle0927 — APPROVED** [review]

**Pull request merged**